src/facade/dragonfly_connection.cc

```diff
@@ -2161,6 +2180,161 @@ bool Connection::WeakRef::operator==(const WeakRef& other) const {
   return client_id_ == other.client_id_;
 }
 
+void Connection::DoReadOnRecv(const util::FiberSocketBase::RecvNotification& n) {
+  if (std::holds_alternative<std::error_code>(n.read_result)) {
+    io_ec_ = std::get<std::error_code>(n.read_result);
+    return;
+  }
+
+  // TODO non epoll API via EnableRecvMultishot
+  // if (std::holds_alternative<io::MutableBytes>(n.read_result))
+  using RecvNot = util::FiberSocketBase::RecvNotification::RecvCompletion;
+  if (std::holds_alternative<RecvNot>(n.read_result)) {
+    if (!std::get<bool>(n.read_result)) {
+      io_ec_ = make_error_code(errc::connection_aborted);
+      return;
+    }
+
+    if (io_buf_.AppendLen() == 0) {
+      // We will regrow in IoLoop
+      return;
+    }
+
+    io::MutableBytes buf = io_buf_.AppendBuffer();
+    io::Result<size_t> res = socket_->TryRecv(buf);
+
+    // error path
+    if (!res) {
+      auto ec = res.error();
+      // EAGAIN and EWOULDBLOCK
+      if (ec == errc::resource_unavailable_try_again || ec == errc::operation_would_block) {
+        return;
+      }
+
+      if (ec == errc::connection_aborted || ec == errc::connection_reset) {
+        // The peer can shutdown the connection abruptly.
+        io_ec_ = ec;
+        return;
+      }
+
+      LOG_EVERY_T(ERROR, 10) << "Recv error: " << ec;
+      io_ec_ = ec;
+      return;
+    }
+
+    if (*res == 0) {
+      io_ec_ = make_error_code(errc::connection_aborted);
+      return;
+    }
+    // A recv call can return fewer bytes than requested even if the
+    // socket buffer actually contains enough data to satisfy the full request.
```

Comment:
why do you say this? yes, there are some corner cases but in general, Recv will try to unload as much as possible and fill the destination buffer